mono/metadata/w32process-win32.c

```diff
@@ -209,6 +214,7 @@ process_complete_path (const gunichar2 *appname, gchar **completed)
 	utf8appmemory = g_utf16_to_utf8 (appname, -1, NULL, NULL, NULL);
 	utf8app = process_unquote_application_name (utf8appmemory);
 
+	MONO_ENTER_GC_SAFE;
 	if (g_path_is_absolute (utf8app)) {
 		*completed = process_quote_path (utf8app);
 		result = TRUE;
```

Comment:
Does this do any I/O?

Reply:
The `g_find_program_in_path` part presumably. Could probably reduce the scope to that. Or enlarge it in caller.

Reply:
And `g_file_test`. Maybe those functions should handle it themselves?

Reply:
Hard to say if these thing should be narrow as possible, or broad.
Doing it the next level down catches all callers.
Doing it higher limits the transitions, if that matters.

Reply:
I don't think we should do GC transitions inside the `g_*` functions - it kind of violates layering.  (but we could make coop-safe versions of them in `mono/utils/`)

Reply:
Have to decide if `g_*` really has meaning outside of mono runtime with coop-compatibility, maybe yes, maybe no, and, if they do, if they themselves shouldn't have two variations, vs. visiting/wrapping all the calls.

So, yes, agreed, but the layering purity might not serve any user, or might be slightly onerous to implement, but ok either way.

Reply:
Yeah it hits a couple functions inside of it, and as Martin brought up elsewhere the heavy use of goto makes it important that we be sure the transition is always hit on both parts and because of the way it expands. I opted to just wrap this whole section rather than a subset so I didn't have to keep transitioning back and forth, but I can't hit just the two parts with the blocking functions because of the gotos. I could rewrite the whole thing, but opted for this instead.

Reply:
Looking at feedback on the other part, I'll just wrap the individual blocking calls and transition back and forth a few times.

Reply:
How expensive are those `MONO_ENTER_GC_SAFE` transitions?

In general, I just don't like this goto pattern very much because it's easy for somebody at a later time either add a goto statement from outside the protected block - or a return statement.

Where is this method called from?  Maybe we should consider protecting the caller instead.

Reply:
`goto` is one of my favorite ways to refactor code to avoid duplication and bugs and simplify error handling and flow control.

Some people consider it evil but I find it is really quite noble.
Opinion is strongly mixed on the matter unfortunately. :(

The alternative is a mess of if/else.

Early return is just as good, if you have no resources to cleanup, invariants to restore, or you have C++ destructors. But that doesn't apply here, so `goto`.